### PR TITLE
Remove extra message actions

### PR DESCRIPTION
**Pull request by rbren**
Fixes https://github.com/OpenDevin/OpenDevin/issues/1595

### CODE REVIEW

**Comment by rbren** [comment]
@li-boxuan what's the process for updating regression/integration tests these days?

**Comment by li-boxuan** [comment]
https://github.com/OpenDevin/OpenDevin/tree/main/tests/integration#write-integration-tests

To update an integration test, it is recommended, but not required (or enforced by CI), to remove the existing test artifacts first. Then you could follow the above steps.

**Comment by rbren** [comment]
@li-boxuan I created a quick `regenerate.sh` in this PR that should help

File: opendevin/events/observation/success.py [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+from opendevin.core.schema import ObservationType
+
+from .observation import Observation
+
+
+@dataclass
+class SuccessObservation(Observation):
+    """
+    This data class represents an error encountered by the agent.
```
Comment (enyst) [inline opendevin/events/observation/success.py:11]:
```suggestion
    This data class represents the result of a successful action.
```

Comment (rbren) [reply opendevin/events/observation/success.py]: good catch!

File: agenthub/codeact_agent/codeact_agent.py [hunk 2]
```diff
@@ -151,19 +147,23 @@ def step(self, state: State) -> Action:
...
                 ), f'{prev_action.__class__} is not supported (supported: {self.SUPPORTED_ACTIONS})'
-                # prev_action is already added to self.messages when returned
+                if (
+                    isinstance(prev_action, MessageAction)
+                    and prev_action.source == 'user'
+                ):
+                    self.messages.append(
+                        {'role': 'user', 'content': prev_action.content}
+                    )
+                    if obs.content.strip() == '/exit':
```
Comment (enyst) [inline agenthub/codeact_agent/codeact_agent.py:157]:
```suggestion
                    if prev_action.content.strip() == '/exit':
```

**Review by enyst — APPROVED** [review]

File: agenthub/codeact_agent/codeact_agent.py (hunk 2 above)
Comment (enyst) [reply agenthub/codeact_agent/codeact_agent.py:157]: This one is necessary, obs is null obs now.

Comment (rbren) [reply agenthub/codeact_agent/codeact_agent.py]: Whoops, another good catch!

**Comment by enyst** [comment]
@rbren 
There is something that we will want to fix: it doesn't display the user messages, in console, it just skips, resulting in a... talk to itself. The monologue strikes back! 😅 

E.g.
```
STEP 0

22:30:43 - PLAN
make a bash script to solve the 17 queens problem
22:31:34 - ACTION
MessageAction(content='Sure, I can help with that. The N-Queens problem is the problem of placing N chess queens on an N×N chessboard so that no two queens threaten each other. Thus, a solution requires that no two queens share the same row, column, or diagonal.\n\nThe 17-Queens problem is a specific case of the N-Queens problem where N is 17. This problem is computationally expensive to solve, especially with a bash script, due to the large number of possibilities to check.\n\nHowever, I can provide a simple bash script that uses a backtracking algorithm to solve the N-Queens problem. This script will not be efficient for large N like 17, but it can solve the problem for smaller N like 8.\n\nHere is the bash script:\n\n```bash\n#!/bin/bash\n\n# N-Queens problem solver in Bash\n\nN=$1\nsolution=0\n\n# Initialize the board\nfor (( i=0; i<$N; i++ )); do\n  for (( j=0; j<$N; j++ )); do\n    board[$i,$j]=0\n  done\ndone\n\n# Check if a position is safe\nis_safe() {\n  local row=$1\n  local col=$2\n  local i=0\n  local j=0\n\n  # Check this row on left side\n  for (( i=0; i<$col; i++ )); do\n    if (( board[$row,$i] )); then\n      return 1\n    fi\n  done\n\n  # Check upper diagonal on left side\n  for (( i=$row, j=$col; i>=0 && j>=0; i--, j-- )); do\n    if (( board[$i,$j] )); then\n      return 1\n    fi\n  done\n\n  # Check lower diagonal on left side\n  for (( i=$row, j=$col; j>=0 && i<$N; i++, j-- )); do\n    if (( board[$i,$j] )); then\n      return 1\n    fi\n  done\n\n  return 0\n}\n\n# A recursive utility function to solve N-Queens problem\nsolve_n_queens() {\n  local col=$1\n  if (( $col == $N )); then\n    (( solution++ ))\n    echo "Solution $solution:"\n    print_solution\n    return\n  fi\n  for (( i=0; i<$N; i++ )); do\n    if is_safe $i $col; then\n      board[$i,$col]=1\n      solve_n_queens $(( $col + 1 ))\n      board[$i,$col]=0\n    fi\n  done\n}\n\n# Function to print a solution\nprint_solution() {\n  for (( i=0; i<$N; i++ )); do\n    for (( j=0; j<$N; j++ )); do\n      echo -n "${board[$i,$j]} "\n    done\n    echo\n  done\n  echo\n}\n\n# Solve the N-Queens problem\nsolve_n_queens 0\n```\n\nYou can run this script with the number of queens as an argument, like this:\n\n```bash\nbash n_queens.sh 8\n```\n\nThis will print all solutions to the 8-Queens problem.\n\nPlease note that this script uses a simple backtracking algorithm and is not optimized for large N. It may take a very long time to find all solutions for N=17.', wait_for_response=True, action='message')

==============
STEP 1

22:34:05 - ACTION
MessageAction(content="I'm sorry for the confusion, but as an AI text-based model, I don't have the capability to execute bash scripts directly. However, you can run the script on your local machine by following these steps:\n\n1. Copy the script provided in the previous response.\n2. Open a text editor, paste the script, and save it as `n_queens.sh`.\n3. Open a terminal and navigate to the directory where you saved `n_queens.sh`.\n4. Make the script executable by running `chmod +x n_queens.sh`.\n5. Run the script with the number of queens as an argument, like this: `./n_queens.sh 3`.\n\nPlease note that the N-Queens problem has no solutions for N=2 and N=3. The first non-trivial case, with a solution, is N=4.", wait_for_response=True, action='message')

==============
STEP 2

22:36:26 - ACTION
**CmdRunAction**
THOUGHT:I apologize for the confusion. You're right, I can execute bash commands. However, creating and executing a bash script in this environment is a bit tricky. Let's try it step by step.
```

**Comment by li-boxuan** [comment]
> I created a quick `regenerate.sh` in this PR that should help

Thanks @rbren . I was experimenting with some more automatic approaches, e.g. let the pytest decide (based on an env variable) whether it should use mock LLM (for testing) or real LLM (to generate results). But it was very cumbersome and didn't go well. I think the script you made is handy enough.

**Comment by rbren** [comment]
Oops @enyst I just saw this message, looks like auto-merge beat me.

Are you seeing this behavior on main? monologue has been working for me

**Comment by enyst** [comment]
@rbren I will be able to check it again in ~8 hours ish. I'm not sure if it was fixed.

I _think_ it was CodeAct agent, not monologue agent; running with UI, the UI did display both sides of course, but the parallel console display was one-sided.

I don't know how I somehow forgot this one these days, crazy days but still, it was incredibly amusing in a way, CodeAct's own attempt at a monologue!

**Comment by enyst** [comment]
https://github.com/OpenDevin/OpenDevin/issues/1736

**Pull request merged**